mergify_engine/gitter.py

```diff
@@ -0,0 +1,96 @@
+#
+# Copyright © 2021 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import asyncio
+import dataclasses
+import logging
+import shutil
+import tempfile
+import typing
+import urllib.parse
+
+from mergify_engine import config
+
+
+@dataclasses.dataclass
+class GitError(Exception):
+    returncode: int
+    output: str
+
+
+@dataclasses.dataclass
+class Gitter(object):
+    logger: logging.LoggerAdapter
+    tmp: str = dataclasses.field(init=False)
+
+    # Worker timeout at 5 minutes, so ensure subprocess return before
+    GIT_COMMAND_TIMEOUT: int = dataclasses.field(init=False, default=4 * 60 + 30)
+
+    async def init(self) -> None:
+        self.tmp = await asyncio.to_thread(tempfile.mkdtemp, prefix="mergify-gitter")
+        self.logger.info("working in: %s", self.tmp)
+        await self("init")
+
+    async def __call__(self, *args: str, input: typing.Optional[str] = None) -> str:
+        self.logger.info("calling: %s", " ".join(args))
+        try:
+            p = await asyncio.create_subprocess_exec(
+                "git",
+                *args,
+                cwd=self.tmp,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.STDOUT,
+                stdin=None if input is None else asyncio.subprocess.PIPE,
+            )
+
+            stdout, _ = await asyncio.wait_for(
+                p.communicate(input=None if input is None else input.encode("utf8")),
+                self.GIT_COMMAND_TIMEOUT,
+            )
```

Comment:
What happens if that raises `TimeoutError`?

Reply:
Before it was raising `TimeoutExpired` and it was not handled, and worker just retry it later. It should have been logged in sentry, if that happen. I just keep the same behavior.